### PR TITLE
Keep shader graph properties order in particle output

### PR DESCRIPTION
**Pull request by julienamsellem**
### Purpose of this PR
https://fogbugz.unity3d.com/f/cases/1398256/

When a shader graph is used in a particle output block the shader graph exposed properties are sorted in the creation order instead of the same order as in the shader graph blackboard.

[Before]
VFX_UX_SG_PropertiesOrder

[After]
Unity_lvPrFf1wkh

---
### Testing status
Tested all output particle blocks with a shader graph property.

---
### Comments to reviewers
If you change the properties order in the shader graph blackboard you must save the shader graph asset so that it's taken into account in the VFX graph.

### CODE REVIEW

**Review by MarieGuff — APPROVED** [review]
No issue found, great improvments!

**Comment by VitaVFX** [comment]
Hey Julien! Thanks for addressing the issue! Here's the WIP [test doc](https://docs.google.com/document/d/196X_TUHa_0u8RN-2NmjN-D4f-EAMTLEJE80eUP4sp2w/edit?usp=sharing) with some issues found🐛 and one request 🤞🏻

File: com.unity.visualeffectgraph/CHANGELOG.md [hunk 1]
```diff
@@ -26,6 +26,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Resize custom operator (multiply, add...) to the minimum size when changing input types
 - Show opened VFX asset in inspector when clicking in a void area and unselect node when VFX graph loose focus
 - Disabled text inputs were unreadable [Case 1387237](https://issuetracker.unity3d.com/product/unity/issues/guid/1387237/)
+- In `Output Particle` blocks, shader graph exposed properties order is consistent with shader graph blackboard
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/CHANGELOG.md:29]: Not sure why this case is public but you can link it here: https://issuetracker.unity3d.com/product/unity/issues/guid/1398256/

File: com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitSphereOutput.cs [hunk 2]
```diff
@@ -57,11 +57,23 @@ protected override IEnumerable<string> filteredOutSettings
...
+
+        protected override IEnumerable<string> untransferableSettings
+        {
+            get
+            {
+                foreach (var setting in base.untransferableSettings)
+                {
+                    yield return setting;
+                }
+                yield return nameof(shaderGraph);
```
Comment (PaulDemeulenaere) [inline com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitSphereOutput.cs:76]: Ok ⏫ We discussed about this, the "Convert Output" was behaving weirdly, the untransferableSettings prevent the creation of Lit or Cube Sphere output 🟢

**Review by PaulDemeulenaere — COMMENTED** [review]
Poke me once you resolve the QA issue to process the last code validation, otherwise, the approach looks correct to me.

File: com.unity.visualeffectgraph/Editor/GraphView/Views/Properties/ObjectPropertyRM.cs [hunk 3]
```diff
@@ -64,9 +65,17 @@ public override void UpdateGUI(bool force)
...
             {
                 m_Value = (UnityObject)obj;
-                m_ValueIcon.image = obj != null
-                    ? AssetPreview.GetMiniTypeThumbnail(m_Value)
-                    : AssetPreview.GetMiniTypeThumbnail(m_Provider.portType);
+                if (m_Provider.portType == typeof(ShaderGraphVfxAsset))
+                {
+                    //m_ValueIcon.image = AssetPreview.GetMiniTypeThumbnail(typeof(Shader));
+                    m_ValueIcon.image = Resources.Load<Texture2D>("Icons/sg_graph_icon");
+                }
```
Comment (julienamsellem) [inline com.unity.visualeffectgraph/Editor/GraphView/Views/Properties/ObjectPropertyRM.cs:72]: This is a hack to display the proper object icon for shader graph. But ideally it should work the same way as other types of objects.
I asked [here](https://unity.slack.com/archives/C47JPNNTZ/p1643731047040519) on slack for help, but I got no answer, so I decided to workaround.

File: com.unity.visualeffectgraph/CHANGELOG.md (hunk 1 above)
Comment (julienamsellem) [reply com.unity.visualeffectgraph/CHANGELOG.md:29]: The link seems broken ...

**Review by esmelusina — APPROVED** [review]
Looks fine from SG.

File: com.unity.visualeffectgraph/CHANGELOG.md (hunk 1 above)
Comment (PaulDemeulenaere) [reply com.unity.visualeffectgraph/CHANGELOG.md:29]: Strange, it's marked as public here https://fogbugz.unity3d.com/f/cases/1398256/

File: com.unity.visualeffectgraph/Editor/ShaderGraph/VFXShaderGraphParticleOutput.cs [hunk 4]
```diff
@@ -235,10 +235,27 @@ public override void SetupMaterial(Material material)
...
+                var properties = ShaderUtil.GetMaterialProperties(new[] { material });
+                foreach (var property in shaderGraph.properties)
+                {
+                    var orderedProperty = properties.SingleOrDefault(x => x.name == property.referenceName);
+                    if (orderedProperty != null)
+                    {
+                        var order = Array.IndexOf(properties, orderedProperty);
+                        if (order != property.order)
+                        {
+                            property.order = order;
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/Editor/ShaderGraph/VFXShaderGraphParticleOutput.cs:248]: If we need an other information in AbstractShaderProperty, we should fill them in [ShaderGraphImporter](https://github.com/Unity-Technologies/Graphics/blob/master/com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs#L500), if not possible, it should a transient data in VFXShaderGraphParticleOutput.
We aren't suppose to modify the `ShaderGraphVfxAsset` in VFX, this is the ownership of shaderGraph.
Investigating on a better solution 👀

Comment (julienamsellem) [reply com.unity.visualeffectgraph/Editor/ShaderGraph/VFXShaderGraphParticleOutput.cs]: Our investigation led to a completely different approach see next commit

File: com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitCubeOutput.cs [hunk 5]
```diff
@@ -52,8 +52,20 @@ protected override IEnumerable<string> filteredOutSettings
...
+
+        protected override IEnumerable<string> untransferableSettings
+        {
+            get
+            {
+                foreach (var setting in base.untransferableSettings)
+                {
+                    yield return setting;
+                }
+                yield return nameof(shaderGraph);
```
Comment (julienamsellem) [inline com.unity.render-pipelines.high-definition/Editor/VFXGraph/Outputs/VFXLitCubeOutput.cs:68]: This change is to fix a different issue, keeping exposed property even after converting an output to a type of output that does not support shadergraph.

**Comment by VitaVFX** [comment]
All the raised issues were addressed, couldn't find anything obscure in the last testing pass. [Test Document](https://docs.google.com/document/d/196X_TUHa_0u8RN-2NmjN-D4f-EAMTLEJE80eUP4sp2w/edit)🟢 

@PaulDemeulenaere My duty is done here ⚔️ you can hop on code review.

**Review by VitaVFX — APPROVED** [review]

File: com.unity.visualeffectgraph/Editor/GraphView/Views/Properties/CustomObjectPicker.cs [hunk 6]
```diff
@@ -62,7 +62,7 @@ private static IEnumerable<SearchItem> FetchShaderGraph(SearchContext context)
             {
                 foreach (var r in request)
                 {
-                    if (r?.ToObject<ShaderGraphVfxAsset>() == null)
+                    if (r != null && r.ToObject<ShaderGraphVfxAsset>() == null)
```
Comment (julienamsellem) [inline com.unity.visualeffectgraph/Editor/GraphView/Views/Properties/CustomObjectPicker.cs:65]: Fixed an issue when opening the object picker the first time

File: com.unity.visualeffectgraph/Editor/GraphView/VFXViewWindow.cs [hunk 7]
```diff
@@ -122,7 +122,7 @@ static VFXViewWindow GetWindowLambda(Func<VFXViewWindow, bool> func, bool create
 
         public static bool CloseIfNotLast(VFXView vfxView)
         {
-            var noAssetWindows = s_VFXWindows.Where(x => x.graphView.controller?.graph == null).ToArray();
+            var noAssetWindows = s_VFXWindows.Where(x => x.graphView?.controller?.graph == null).ToArray();
```
Comment (julienamsellem) [inline com.unity.visualeffectgraph/Editor/GraphView/VFXViewWindow.cs:125]: Fixed an issue that could occur when closing Unity

File: com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs [hunk 8]
```diff
@@ -850,13 +850,16 @@ void AddCoordinateSpaceSnippets(InterpolatorType interpolatorType, Func<ShaderGr
...
-                {
-                    continue;
-                }
+            // Fetch properties from the categories to keep the same order as in the shader graph blackboard
+            // Union with the flat properties collection because previous shader graph version could store properties without category
+            var sortedProperties = graph.categories
+                .SelectMany(x => x.Children)
+                .OfType<AbstractShaderProperty>()
+                .Union(graph.properties)
+                .Where(x => x.isExposed);
```
[the remark this answers is not in the document]
Comment (julienamsellem) [reply com.unity.shadergraph/Editor/Importers/ShaderGraphImporter.cs:859]: Good point, I added a test that uses an old version shader graph with a property which is not in a any category

**Review by PaulDemeulenaere — APPROVED** [review]
Please be sure this [job](https://unity-ci.cds.internal.unity3d.com/job/11664152/logs) doesn't add new failure (see [master](https://unity-ci.cds.internal.unity3d.com/project/902/branch/master/jobDefinition/.yamato%2Fvfx_hdrp-win-dx11.yml%23VFX_HDRP_Win_DX11_editmode_mono_Linear_trunk) for reference)
Otherwise, looks good to me, we already did some pair developement on this PR.
The added editor test are very welcomed.
Thanks for the improvement !

**Comment by PaulDemeulenaere** [comment]
_green
Ready to merge after a couple of successful [edit mode test](https://unity-ci.cds.internal.unity3d.com/project/902/branch/vfx%2Ffix%2F1398256-shadergraph-properties-order/jobDefinition/.yamato%2Fvfx_hdrp-win-dx11.yml%23VFX_HDRP_Win_DX11_editmode_mono_Linear_trunk)

**Pull request merged**